### ISSUE TITLE
Getter and setter functions do not set tag status on errors.

### ISSUE THREAD

**Issue by kyle-github** [libplctag/libplctag#205]
In most cases (`plc_tag_get/set_bit()` is an exception) using the return value to signal an error is insufficient.   There is no current way to determine the error.   For instance if the tag did not finish a read (or failed the initial read), then it may not have a size or any data at all.  You cannot tell that from the response.

A proposal here would be to set the tag status when setting the return value.   This would allow for logic like:

```c
uint32 val = plc_tag_get_uint32(a_tag, 10);
if(val == UINT32_MAX) {
   int rc = plc_tag_status(a_tag);
   if(rc != PLCTAG_STATUS_OK) {
      printf("Got error trying to get value from tag! Error %s!\n", plc_tag_decode_error(rc));
      return rc;
   }
}
/* ... do something with the value ... */
```
This effectively breaks the API by changing which functions change the tag status.

Another alternative would be to add to/change the API to allow for default values:

```c
uint32_t val = plc_tag_get_uint32_ex(a_tag, 10, 0xFFF);
if(val == 0xFFF) {
   printf("Error getting value from the tag!\n");
   return plc_tag_status(a_tag);
}
/* ... do something with the value ... */
```

That has the advantage that the application can decide what value is "invalid" and thus what signals a problem.   It also means only these new getter functions change the tag status, not the old ones and therefore does not break the implicit API.

**Comment by kyle-github** [comment]
Note that the documentation says that these getter functions will set the tag status.   So for now that will be the way we will do things.   Breaking the API should only be done if there is absolutely no other way to fix a problem.   Adding to the API as in the above is a cleaner way to do this but leads to API explosion.   Perhaps in 2.3 or 3.0?

**Comment by kyle-github** [comment]
Also updated to have setters set the tag status.

The status was set in earlier versions of the library but appears to have regressed around the time the Modbus code was added and the getters/setters were refactored.

**Comment by kyle-github** [comment]
Fixed in 2.1.20.
